README
PROMS

	Version     : %%VERSION
	Copyright   : Copyright 2003-2005 by Ferry Boender <f.boender AT electricmonk.nl>.

	Authors     : Ferry Boender <f.boender AT electricmonk.nl> 
	              Michiel van Baak <michiel (at) vanbaak (dot) info>
	License     : General Public License


	PROMS is a webbased project management system whose primary goal is to
	provide a facility for developers to communicate amongst other 
	developers, communicating things to end-users as well as getting feed-
	back from users (in the form of bugs, etc) and as a placeholder for
	project information.


Table of contents

	1. About PROMS
		1.2. Current features
	2. Installation / Upgrade
		2.1. Requirements
		2.2. Installation / Upgrade
			2.2.1. Automated installation / Upgrade
			2.2.2. Manual installation / Upgrade
		2.3. Configuring
		2.4. Setting up the noreply PROMS e-mail address
	3. Feedback / Contributing.
		3.1. Feedback (bugs/questions/feature requests)
		3.2. Contributing
			3.2.1. Writing patches/contributions.
		3.3. Credits
	4. Roadmap
	5. Legal stuff
	6. Authors

1. About PROMS

	PROMS, which stands for Project Management System, is a webbased 
	application which focuses on the management of projects. It's main focus
	is to provide a facility for developers to communicate with other 
	developers and for them to exchange information with end-users.

	A good way of describing PROMS would be 'a combination of Freshmeat and
	SourceForge'.  It provides features for communicating and maintaining
	projects, releases, bugs, forums, files and todo's.
	
	At the moment, PROMS is mostly oriented towards small projects with 
	just a handfull of developers. Therefor things like e-mail notification
	are not really tweaked towards many developers on a single project. 


1.2. Current features.

	PROMS currently includes these features:
	
	* Projects
	* Project parts
	* Bugs / Bugtracker)
	* Todos
	* Files
	* Releases (including subscriber notifications)
	* General discussions
	* Bug discussions
	* Todo discussions
	* User management (including per-user restrictions/rights)
	* User manual (semi-context sensitive)
	* E-mail notifications of changes for most of the above features
	
	For a full list of all features accompanied by descriptions, please refer
	to the website. 
	
	PROMS is currently under development, and it is not recommended for daily
	usage. (Though it has never deleted any of my personal project stuff)
	
2. Installation

2.1. Requirements

	Serverside requirements:
		* PHP 4.x (might work under 3.x)
		* MySQL 3.x or greater.

	Clientside requirements:
		* Javascript enabled browser (only basic javascript like window.open 
		  and navigation.back)
		* CSS2 (will work without, but looks horrible)

	Developed and tested using:
		* Mozilla Firebird 0.6 - 1.0 / Internet Explorer 5/6
		* Apache 1.3.26
		* PHP 4.1.2
		* MySQL 3.23.49
		
2.2. Installation / Upgrade

	2.2.1. Automated installation / Upgrade
	
	Since version 0.9, PROMS contains an automated installer. This installer
	will guide you through the setup step by step. It will create a database
	for you and set up and install the sourcecode for you if you provide it
	with some basic information. It can also upgrade your version of PROMS from
	any version to the current version. 

	You can start the installer by executing the   setup.sh   script. If you
	follow the instructions on the screen you should be just fine. 

	The setup script can be run in two ways:

	[root@localhost] ~# chmod 755 setup.sh; ./setup.sh

	or

	[root@localhost] ~# sh ./setup.sh


	Please remember that the setup script is a beta (just like the rest of
	PROMS at this time). It might screw up your database. The setup script
	creates a backup of your database before upgrading, or at least it tries.
	If anything goes wrong, you should be able to put the old database back
	using this backup.
	
	NOTICES: 
		Please make sure you have have Magic Quotes enabled in your
		installation of PHP. If you do not have this enabled, your installation
		of PROMS will be VULNERABLE to SQL INJECTIONS!

		PROMS requires that the 'short_open_tag' be set to 'on' in the php.ini
		configuration.

		If you are running a different web server than Apache, please use your
		web server's configuration to prevent people from running scripts in 
		the files/ directory! If you do not, people might be able to upload PHP
		or PERL files and execute from within PROMS, which opens up major
		security holes.

	2.2.2. Manual installation / Upgrade

	If, for some reason, the automated installer does not work for you, you can
	install PROMS manually. (Please report any problems with the automated 
	installer!). 

	To install a completely new version, follow these steps:

	1. Create a new MySQL database.
	2. Import the file sql/proms.sql into the database:
	   [todsah@sharky]~$ mysql -u todsah -p [databasename] < sql/proms.sql
	3. Create a new user for the database:
	   mysql> GRANT ALL PRIVILEGES ON [databasename].* TO proms@localhost 
       IDENTIFIED BY '[password]';
	4. Change the defines in the src/settings.php file to match with the 
	   database and user names. Also change the other settings to your liking.
	   Remove the 'REF00#_' from each of the defines so that the lines look 
	   like this: : define ("DB_DATABASE", "proms");

	   ! Make sure you set the PROMS_VERSION variable to the correct version.
	5. Copy all .php files in the src/ directory to your webserver directory.
	6. Point your browser to the proms.php page.
	7. Register a new account for yourself. The first account registered will
	   be the administrator account whom can do anything he damn well pleases.
	   

	To upgrade a current version:

	1. Run the scripts in the sql/ directory. You should check your current
	   version of PROMS and run each script with a version number equal to, or
	   higher than, your current version. Running the scripts can be done 
	   with the following command:

	   mysql -uUSERNAME -pPASSWORD -hHOSTNAME DATABASENAME < sql/0.8.sql

	   Of course you must replace USERNAME et al with the username of a mysql-
	   account which is allowed to create/mod tables and such.
	2. Change the defines in the src/settings.php file to match with the
	   database and usernames. Also check all other settings and set them to
	   their appropriate values (shouldn't be too difficult).
	3. Copy src/* to the destination path.
	

	NOTICES: 
		Please make sure you have have Magic Quotes enabled in your
		installation of PHP. If you do not have this enabled, your installation
		of PROMS will be VULNERABLE to SQL INJECTIONS!


CHANGELOG
Proms: ChangeLog
------------------------------------------------------------------------------

[0.12.1] <f.boender AT electricmonk.nl> ( Fri Oct 21 11:21:00 CET 2005 )

  * Fixed a bug in the setup script which caused upgrades to fail.

[0.12] <f.boender AT electricmonk.nl> ( Fri Oct 21 11:08:00 CET 2005 )

  * Fixed a bug in the mail sending routine that caused a failure in the mailing of notifications.
  * Fixed a bug in the permissions while creating the directories that hold a project's files.

  * A notice was added to the README documenting the fact that PHP's short_open_tag setting must be on.
  * The 'Manage' tab is no longer shown if the user is not logged in.
  * Projects can now be flagged as 'private' so that they will only show up for project members and the project owner.
  * Added the ability to delete todo's.
  * Added the ability to delete files.

[0.11] <f.boender AT electricmonk.nl> ( Thu May 5 08:19:00 CET 2005 )

  Bug^H^H^Hfeature fixes

  * Various SQL queries where vulnerable to SQL injections. Fixed. (See also
    README)
  * Various modules (Project Parts, Project Members, Todo, Discussions, ..) 
    had incorrect error checking. Fixed.
  * It wasn't possible to view discussions if not logged in. Fixed.
  * Certain combinations of filters in the Bug list didn't work. Fixed.
  * Certain combinations of filters in the Todo list didn't work. Fixed.
  * Some typo fixes.
  * In certain cases it wasn't possible to update bugreport details. Fixed.
  * Wrong bug number was mentioned in email notifications of new bugreports. 
    Fixed.
  * A few places didn't filter out HTML entities correctly. Fixed.
  * In some posts in the discussion forums no username would be shown for 
    anonymous users. Fixed.
  * When viewing a bug or todo discussion, navigating back would take you to 
    the discussion list instead of the bug or todo. Fixed.
  * Todo's used to use their database key as an id, resulting in todo nr's that
    didn't start at 1 for each project. Fixed.
  * Certain combinations of rights caused users to be granted more rights then
    they should have been. Fixed.
  * Various list filters were to wide for 1024x768. Fixed.
  * 'New Project' wizard didn't work. Fixed.
  * Various improvements in the security checks. Many checks depended on 
    being the project owner where they should have depended on the individual
	access right. Fixed.
  * Bug and todo notification email URLs always pointed to http://, but this 
    could also be https://. Fixed.
  * It was possible for non-authorized users to view and modify the list of 
    project members. Fixed.
  * 'New project' wizard was broken. Fixed.
  * Todos could be modified by non-authorized users. Fixed.

  New features / Improvements.

  * Replying to a message in a discussion now shows the parent post.
  * Better email notifications of new posts in a discussion.
  * Project parts now can have their own maintainer.
  * Major interface changes. Tabs are now used for the various parts of a 
    project. 
  * New generic file management interface. 
  * Coupling between file management and Project releases.
  * Discussions: 'Create new thread' and 'Post reply' buttons are now not shown
    anymore when not logged in.

setup.sh
#!/bin/sh
#
# PROMS setup/installation tool
#
# Copyright (C), 2003-2005 Ferry Boender. Released under the General Public
# License For more information, see the COPYING file supplied with this
# program.
#

#----------------------------------------------------------------------------
# input and output functions
#----------------------------------------------------------------------------
function input_line() {
	if [ "$HAS_DIALOG" == "1" ]; then
		LINE=`dialog --no-cancel --title "${1}" --stdout --inputbox "${2}" 0 0 "${4}"`;
	else
		clear
		echo -e "${1}\n---------------------------------------------------------------------------\n"
		if [ "$HAS_FMT" == "1" ]; then
			echo -e "${2}\n" | fmt
		else
			echo -e "${2}\n"
		fi
		read -p"[${4}]" LINE
		if [ -z $LINE ]; then
			LINE=${4}
		fi
	fi

	eval $3='$LINE'
}

function input_password() {
	if [ "$HAS_DIALOG" == "1" ]; then
		LINE=`dialog --no-cancel --title "${1}" --stdout --passwordbox "${2}" 0 0 "${4}"`;
		read
	else
		clear
		echo -e "${1}\n---------------------------------------------------------------------------\n"
		if [ "$HAS_FMT" == "1" ]; then
			echo -e "${2}\n" | fmt
		else
			echo -e "${2}\n"
		fi
		read -s -p"[${4}]" LINE
		if [ -z $LINE ]; then
			LINE=${4}
		fi
	fi
	
	eval $3='$LINE'
}

function input_yesno() {
	if [ "$HAS_DIALOG" == "1" ]; then
		dialog --title "${1}" --yesno "${2}" 0 0;
		eval $3='$?';
	else
		clear
		echo -e "${1}\n---------------------------------------------------------------------------\n"
		if [ "$HAS_FMT" == "1" ]; then
			echo -e "${2}\n" | fmt
		else
			echo -e "${2}\n"
		fi
		read -n1 -p"[Y/N]" YESNO
		echo -e "\n"
		if [ "$YESNO" == "n" -o "$YESNO" == "N" ]; then
			eval $3='1';
		else
			eval $3='0';
		fi
	fi
}

function output_text() {
	if [ "$HAS_DIALOG" == "1" ]; then
		dialog --title "${1}" --msgbox "$2" 0 0;
	else 
		clear
		echo -e "${1}\n---------------------------------------------------------------------------\n"
		if [ "$HAS_FMT" == "1" ]; then
			echo -e "${2}\n" | fmt
		else
			echo -e "${2}\n"
		fi
		read -p"-- press enter to continue --"
	fi
}

#----------------------------------------------------------------------------
# Menus
#----------------------------------------------------------------------------
function actions_menu() {
	TMP_FILE=`tempfile`

	DONE_TMP=$DONE
	DONE=0
	while [ "$DONE" == "0" ]; do # Settings check 
		while [ "$DONE" == "0" ]; do
			if [ "$HAS_DIALOG" == "1" ]; then
				dialog --ok-label Select --title "PROMS Setup" --menu "Please choose an action" 22 60 15 \
					1 "Read settings from previous installation" \
					2 "View / Change settings" \
					3 "Change database and webserver settings for setup" \
					0 "Done (start installation/upgrade)" 2>$TMP_FILE
				if [ "$?" == "1" ]; then
					rm $TMP_FILE
					exit
				else
					ACTION=`cat $TMP_FILE`
				fi

			else
				clear
				echo -e "PROMS Setup\n---------------------------------------------------------------------------\n${2}\n"
				echo "1 - Read settings from previous installation"
				echo "2 - View / Change settings"
				echo "3 - Change database and webserver settings for setup"
				echo
				echo "0 - Done (start installation)"
				echo "C - Cancel (stop installation)"
				echo
				read -p"Choice: " ACTION
			fi

			case "$ACTION" in
				1) settings_get_upgrade ;;
				2) settings_change_menu ;;
				3) setup_db_change_menu ;;
				C) rm $TMP_FILE; exit ;;
				c) rm $TMP_FILE; exit ;;
				0) ACTION=""; DONE=1 ;;
			esac

		done

		settings_check # Sets DONE to 0 if some settings are missing.
	done

	DONE=$DONE_TMP

	rm $TMP_FILE
}

function settings_change_menu() {
	TMP_FILE=`tempfile`

	DONE_TMP=$DONE
	DONE=0

	while [ "$DONE" == "0" ]; do
		if [ "$HAS_DIALOG" == "1" ]; then
			dialog --cancel-label Done --ok-label Change --title "Change settings" --menu "Please choose a setting to change" 22 60 16 \
				"Install location" "$S_SETUP_PATH" \
				"DB hostname" "$S_PROMS_DB_HOSTNAME" \
				"DB username" "$S_PROMS_DB_USERNAME" \
				"DB password" "$S_PROMS_DB_PASSWORD"\
				"DB name" "$S_PROMS_DB_NAME"\
				"E-mail address" "$S_PROMS_LIST_EMAIL" \
				"SMTP hostname" "$S_PROMS_SMTP_HOSTNAME" \
				"SMTP port" "$S_PROMS_SMTP_PORT" \
				2>$TMP_FILE
			if [ "$?" == "1" ]; then
				DONE=1
				ACTION=""
			else 
				ACTION=`cat $TMP_FILE`
			fi
			
		else
			clear
			echo -e "Change settings\n---------------------------------------------------------------------------\n${2}\n"
			echo "1 - Install location : $S_SETUP_PATH"
			echo "2 - DB hostname      : $S_PROMS_DB_HOSTNAME"
			echo "3 - DB username      : $S_PROMS_DB_USERNAME"
			echo "4 - DB password      : $S_PROMS_DB_PASSWORD"
			echo "5 - DB name          : $S_PROMS_DB_NAME"
			echo "6 - E-mail address   : $S_PROMS_LIST_EMAIL"
			echo "7 - SMTP hostname    : $S_PROMS_SMTP_HOSTNAME"
			echo "8 - SMTP port        : $S_PROMS_SMTP_PORT"
			echo
			echo "0 - Done"
			echo
			read -p"Choice: " ACTION
			
			case "$ACTION" in
				1) ACTION="Install location";;
				2) ACTION="DB hostname"     ;;
				3) ACTION="DB username"     ;;
				4) ACTION="DB password"     ;;
				5) ACTION="DB name"         ;;
				6) ACTION="E-mail address"  ;;
				7) ACTION="SMTP hostname"   ;;
				8) ACTION="SMTP port"       ;;
				0) ACTION=""; DONE=1 ;;
			esac
		fi

		case "$ACTION" in
			"Install location" ) settings_change_installpath ;;
			"DB hostname"      ) settings_change_db_hostname ;;
			"DB username"      ) settings_change_db_username ;;
			"DB password"      ) settings_change_db_password ;;
			"DB name"          ) settings_change_db_name ;;
			"E-mail address"   ) settings_change_email ;;
			"SMTP hostname"    ) settings_change_smtp_hostname ;;
			"SMTP port"        ) settings_change_smtp_port ;;
			*) ;; 
		esac
	done

	DONE=$DONE_TMP

	rm $TMP_FILE
}

function setup_db_change_menu() {
	TMP_FILE=`tempfile`

	DONE_TMP=$DONE
	DONE=0

	while [ "$DONE" == "0" ]; do
		while [ "$DONE" == "0" ]; do
			if [ "$HAS_DIALOG" == "1" ]; then
				dialog --cancel-label Done --ok-label Change --title "Change DB & Web Access" --menu "Please choose a setting to change" 22 60 16 \
					"DB Admin hostname" "$S_SETUP_DB_HOSTNAME" \
					"DB Admin username" "$S_SETUP_DB_USERNAME" \
					"DB Admin password" "$S_SETUP_DB_PASSWORD"\
					"Webserver Username" "$S_SETUP_WEB_USERNAME"\
					"Webserver Groupname" "$S_SETUP_WEB_GROUPNAME"\
					2>$TMP_FILE
				if [ "$?" == "1" ]; then
					DONE=1
					ACTION=""
				else 
					ACTION=`cat $TMP_FILE`
				fi
			else
				clear
				echo -e "Change DB & Web Access\n---------------------------------------------------------------------------\n${2}\n"
				echo "1 - DB Admin hostname   : $S_SETUP_DB_HOSTNAME"
				echo "2 - DB Admin username   : $S_SETUP_DB_USERNAME"
				echo "3 - DB Admin password   : $S_SETUP_DB_PASSWORD"
				echo "4 - Webserver Username  : $S_SETUP_WEB_USERNAME"
				echo "5 - Webserver Groupname : $S_SETUP_WEB_GROUPNAME"
				echo
				echo "0 - Done"
				echo
				read -p"Choice: " ACTION
				case "$ACTION" in
					1) ACTION="DB Admin hostname"   ;;
					2) ACTION="DB Admin username"   ;;
					3) ACTION="DB Admin password"   ;;
					4) ACTION="Webserver Username"  ;;
					5) ACTION="Webserver Groupname" ;;
					0) ACTION=""; DONE=1 ;;
				esac
			fi

			case "$ACTION" in
				"DB Admin hostname"   ) setup_change_db_hostname ;;
				"DB Admin username"   ) setup_change_db_username ;;
				"DB Admin password"   ) setup_change_db_password ;;
				"Webserver Username"  ) setup_change_web_username;;
				"Webserver Groupname" ) setup_change_web_groupname;;
				*) ;; 
			esac
		done

		settings_check_mysql # Modifies $DONE
	done

	DONE=$DONE_TMP

	rm $TMP_FILE
}

#----------------------------------------------------------------------------
# Special function related to settings
#----------------------------------------------------------------------------
function settings_check_mysql() {
	mysql -u"$S_SETUP_DB_USERNAME" --password="$S_SETUP_DB_PASSWORD" -h "$S_SETUP_DB_HOSTNAME" -e"exit"

	if [ "$?" == "1" ]; then
		input_yesno "Database test" "Testing the database connection...\n\nDatabase connection FAILED\n\nDo you want to change the information?" "DONE";
	else
		output_text "Database test" "Testing the database connection...\n\nDatabase connection succesful."
		DONE=1
	fi
}

function settings_check() {
	ERRORS=""
	
	if [ -z "$S_PROMS_DB_HOSTNAME" ]; then
		ERRORS="$ERRORS\nNo database hostname was specified for PROMS to connect to."
	fi
	if [ -z "$S_PROMS_DB_USERNAME" ]; then
		ERRORS="$ERRORS\nNo database username was specified for PROMS to connect with."
	fi
	if [ -z "$S_PROMS_DB_PASSWORD" ]; then
		ERRORS="$ERRORS\nNo database password was specified for PROMS to connect with."
	fi
	if [ -z "$S_PROMS_DB_NAME" ]; then
		ERRORS="$ERRORS\nNo database name was specified for PROMS to connect to."
	fi
	if [ -z "$S_PROMS_SMTP_HOSTNAME" ]; then
		ERRORS="$ERRORS\nNo SMTP hostname was specified for PROMS to sent mail through."
	fi
	if [ -z "$S_PROMS_SMTP_PORT" ]; then
		ERRORS="$ERRORS\nNo SMTP port was specified for PROMS to sent mail through"
	fi
	if [ -z "$S_PROMS_LIST_EMAIL" ]; then
		ERRORS="$ERRORS\nNo E-mail address for PROMS was specified."
	fi

	if [ -z "$S_SETUP_PATH" ]; then
		ERRORS="$ERRORS\nNo installation location was specified."
	fi
	if [ -z "$S_SETUP_DB_HOSTNAME" ]; then
		ERRORS="$ERRORS\nNo hostname was specified for the setup database connection."
	fi
	if [ -z "$S_SETUP_DB_USERNAME" ]; then
		ERRORS="$ERRORS\nNo username was specified for the setup database connection."
	fi
	if [ -z "$S_SETUP_DB_PASSWORD" ]; then
		ERRORS="$ERRORS\nNo password was specified for the setup database connection."
	fi
	if [ -z "$S_SETUP_WEB_USERNAME" ]; then
		ERRORS="$ERRORS\nThe webserver's username was not specified."
	fi
	if [ -z "$S_SETUP_WEB_GROUPNAME" ]; then
		ERRORS="$ERRORS\nThe webserver's groupname was not specified."
	fi

	if [ -n "$ERRORS" ]; then 
		output_text "Incomplete setup" "Some settings are not complete:\n$ERRORS\n\nPlease go back and correct them"
		DONE=0
	fi
}

function settings_get_upgrade() {
	DONE_TMP=$DONE
	DONE=0
		
	while [ "$DONE" == "0" ]; do
		input_line "Previous settings" "If you already have a version of PROMS installed, the settings from that\ninstallation can be read and re-used.\n\nPlease enter the full path where the previous version of PROMS is installed." "S_PREV_PATH" "$S_PREV_PATH"
		S_PREV_PATH=`echo $S_PREV_PATH| sed -e 's/\/$//'`
		S_SETUP_PATH="$S_PREV_PATH"

		if [ -e "$S_PREV_PATH/settings.php" ]; then
			# Get settings from previous settings.php
			S_PREV_VERSION=`cat $S_PREV_PATH/settings.php | grep PROMS_VERSION | grep -v REF| cut -d"\"" -f4`
			S_PROMS_LIST_EMAIL=`cat $S_PREV_PATH/settings.php | grep PROMS_EMAIL | grep -v REF| cut -d"\"" -f4`
			S_PROMS_DB_HOSTNAME=`cat $S_PREV_PATH/settings.php | grep DB_HOSTNAME | grep -v REF| cut -d"\"" -f4`
			S_PROMS_DB_USERNAME=`cat $S_PREV_PATH/settings.php | grep DB_USERNAME | grep -v REF| cut -d"\"" -f4`
			S_PROMS_DB_PASSWORD=`cat $S_PREV_PATH/settings.php | grep DB_PASSWORD | grep -v REF| cut -d"\"" -f4`
			S_PROMS_DB_NAME=`cat $S_PREV_PATH/settings.php | grep DB_DATABASE | grep -v REF| cut -d"\"" -f4`
			S_PROMS_SMTP_HOSTNAME=`cat $S_PREV_PATH/settings.php | grep SMTP_HOST | grep -v REF| cut -d"\"" -f4`
			S_PROMS_SMTP_PORT=`cat $S_PREV_PATH/settings.php | grep SMTP_PORT | grep -v REF| cut -d"\"" -f4`

			# Due to a bug in version 0.11, the version number was removed from settings.php
			# This is a hack to fix that problem.
			if [ -z "$S_PREV_VERSION" ]; then
				S_PREV_VERSION="0.11"
			fi
			
			output_text "Previous settings" "The settings from the previous installation have been read. Please review these settings to see if any need to be changed.";
			DONE="1"
		elif [ -e "$INSTALL_PATH/inc_common.php" ]; then
			# Get settings from previous inc_common.php (PROMS v0.8)
			S_PREV_VERSION="0.8"
			S_PROMS_DB_HOSTNAME=`cat $INSTALL_PATH/inc_common.php | grep "\\\$db_hostname" | grep -v "(" | cut -d"\"" -f2`
			S_PROMS_DB_USERNAME=`cat $INSTALL_PATH/inc_common.php | grep "\\\$db_username" | grep -v "(" | cut -d"\"" -f2`
			S_PROMS_DB_PASSWORD=`cat $INSTALL_PATH/inc_common.php | grep "\\\$db_password" | grep -v "(" | cut -d"\"" -f2`
			S_PROMS_DB_NAME=`cat $INSTALL_PATH/inc_common.php | grep "\\\$db_database" | grep -v "(" | cut -d"\"" -f2`
			# SMTP Stuff wasn't present in 0.8 yet, so we can safely skip this here.

			output_text "Previous settings" "The settings from the previous installation have been read. Please review these settings to see if any need to be changed.";
			DONE="1"
		else 
			input_yesno "Previous settings" "No previous installation of PROMS was found at the path specified. Do you want to specify a different path?" "DONE";
		fi
	done

	DONE=$DONE_TMP
}

function settings_get_defaults() {
	VERSION="%%VERSION"
	S_PREV_VERSION=""
	S_PREV_PATH=""

	S_SETUP_PATH="`grep -ih "documentroot" /etc/apache/* 2>/dev/null | grep -v "^[	 ]*#" | tail -n 1 | sed -e "s/^[	 ]*DocumentRoot[	 ]*//" -e "s/ $//" | sed -e "s/\/$//"`/proms"
	S_SETUP_DB_HOSTNAME="localhost"
	S_SETUP_DB_USERNAME="root"
	S_SETUP_DB_PASSWORD=""
	S_SETUP_WEB_USERNAME="`grep -ih -e "user[	 ]" /etc/apache/* 2>/dev/null | grep -v "^[	 ]*#" | tail -n 1 | sed -e "s/^[	 ]*User[	 ]*//" | sed -e "s/ $//" | sed -e "s/\/$//"`"
	S_SETUP_WEB_GROUPNAME="`grep -ih -e "group[	 ]" /etc/apache/* 2>/dev/null | grep -v "^[	 ]*#" | tail -n 1 | sed -e "s/^[	 ]*Group[	 ]*//" | sed -e "s/ $//" | sed -e "s/\/$//"`"

	S_PROMS_DB_HOSTNAME="localhost"
	S_PROMS_DB_USERNAME="proms"
	S_PROMS_DB_PASSWORD=""
	S_PROMS_DB_NAME="proms"
	S_PROMS_SMTP_HOSTNAME="localhost"
	S_PROMS_SMTP_PORT="25"
	S_PROMS_LIST_EMAIL="proms@`dnsdomainname`"

	if [ "$S_PROMS_LIST_EMAIL" == "proms@" ]; then
		S_PROMS_LIST_EMAIL="proms@`dnsdomainname -f`"
	fi
}

#----------------------------------------------------------------------------
# Setting manipulation functions
#----------------------------------------------------------------------------
function settings_change_installpath() {
	input_line "Installation path" "At which path do you want to install PROMS?" "S_SETUP_PATH" "$S_SETUP_PATH"
}

function settings_change_db_hostname() {
	input_line "Database" "PROMS needs to connect to the database server where PROMS' database is kept. For this it needs some information about the server.\n\nPlease enter the HOSTNAME of the server where the database is running." "S_PROMS_DB_HOSTNAME" "$S_PROMS_DB_HOSTNAME"
}

function settings_change_db_username() {
	input_line "Database" "PROMS needs to connect to the database server where PROMS' database is kept. For this it needs some information about the server.\n\nPlease enter the USERNAME with which you want the PROMS code to connect. " "S_PROMS_DB_USERNAME" "$S_PROMS_DB_USERNAME"
}

function settings_change_db_password() {
	input_line "Database" "PROMS needs to connect to the database server where PROMS' database is kept. For this it needs some information about the server.\n\nPlease enter the PASSWORD with which you want the PROMS code to connect. If you are upgrading, and the previous password was correct, you can simply press enter here." "S_PROMS_DB_PASSWORD" "$S_PROMS_DB_PASSWORD"
}

function settings_change_db_name() {
	input_line "Database" "PROMS needs to connect to the database server where PROMS' database is kept. For this it needs some information about the server.\n\nPlease enter the DATABASE NAME in which you want to put the PROMS data." "S_PROMS_DB_NAME" "$S_PROMS_DB_NAME"
}

function settings_change_email() {
	input_line "Settings" "PROMS has the ability to send out notifications of new bugs, todo's and releases via e-mails. PROMS needs an email address which can be used as a return address as well as the primary e-mail recipient.\n\nPlease check the README to find out how to set up this e-mail account on your system.\n\nPlease enter the e-mail address you want to use for PROMS." "S_PROMS_LIST_EMAIL" "$S_PROMS_LIST_EMAIL"
}

function settings_change_smtp_hostname() {
	input_line "SMTP" "PROMS need as SMTP server to relay email notifications.\n\nPlease enter the HOSTNAME of the SMTP server you want to use for PROMS." "S_PROMS_SMTP_HOSTNAME" "$S_PROMS_SMTP_HOSTNAME"
}

function settings_change_smtp_port() {
	input_line "SMTP" "PROMS need as SMTP server to relay email notifications.\n\nPlease enter the PORT of the SMTP server you want to use for PROMS." "S_PROMS_SMTP_PORT" "$S_PROMS_SMTP_PORT"
}

function setup_change_db_hostname() {
	input_line "DB Admin hostname" "Please enter the HOSTNAME of the database on which PROMS will run." "S_SETUP_DB_HOSTNAME" "$S_SETUP_DB_HOSTNAME"
}

function setup_change_db_username() {
	input_line "DB Admin username" "Please enter the ADMIN USERNAME of the database on which PROMS will run." "S_SETUP_DB_USERNAME" "$S_SETUP_DB_USERNAME"
}

function setup_change_db_password() {
	input_line "DB Admin password" "Please enter the ADMIN PASSWORD of the database on which PROMS will run." "S_SETUP_DB_PASSWORD" "$S_SETUP_DB_PASSWORD"
}

function setup_change_web_username() {
	input_line "Webserver username" "Please enter the Username under which the webserver runs" "S_SETUP_WEB_USERNAME" "$S_SETUP_WEB_USERNAME"
}

function setup_change_web_groupname() {
	input_line "Webserver groupname" "Please enter the Groupname under which the webserver runs" "S_SETUP_WEB_GROUPNAME" "$S_SETUP_WEB_GROUPNAME"
}

#----------------------------------------------------------------------------
# Script initialisation
#----------------------------------------------------------------------------
if [ -x /usr/bin/dialog ]; then
	HAS_DIALOG=1
else
	HAS_DIALOG=0
fi

if [ -x /usr/bin/fmt ]; then
	HAS_FMT=1
else
	HAS_FMT=0
fi

if [ `id -u` != 0 ]; then 
	HAS_ROOT=0
else
	HAS_ROOT=1
fi

output_text "PROMS" "PROMS v$VERSION setup\n\nPROMS is copyright by Ferry Boender. Released under the General Public License (GPL). See the COPYING file for more information.\n\nThis setup script will install PROMS on your system. In order to be of any help with installing, this script will need certain information from you. Please walk through the options in the menu to set this information. If default values are available they will be given.\n\nThis setup tool is UNSTABLE and BETA. It might very well DESTROY your database all together. Make backups.";

if [ "$HAS_ROOT" == "0" ]; then
	input_yesno "Not root" "In order to properly install PROMS you will have to be root. You are not root. You may proceed with this setup script, but you will not be able to use the default installation values.\n\nDo you want to continue?" CONT_NONROOT;
	if [ "$CONT_NONROOT" == "1" ]; then
		exit;
	fi
fi 

settings_get_defaults;
actions_menu

#----------------------------------------------------------------------------
# Deploy / Upgrade database
#----------------------------------------------------------------------------
if [ ! -z $S_PREV_VERSION ]; then
	#
	# Upgrade database
	#
	
	# Create backups before doing anything
	echo "Backing up previous database into backup.sql";
	mysqldump -u$S_SETUP_DB_USERNAME -h$S_SETUP_DB_HOSTNAME --password="$S_SETUP_DB_PASSWORD" $S_PROMS_DB_NAME > backup.sql

	# Upgrade database
	echo "Upgrading database";

	for DB_UPDATE in `grep -x -A1000 "$S_PREV_VERSION" ./sql/db_upgrade_path`; do
		mysql -u$S_SETUP_DB_USERNAME -h$S_SETUP_DB_HOSTNAME --password="$S_SETUP_DB_PASSWORD" $S_PROMS_DB_NAME < sql/$DB_UPDATE.sql
	done
else
	# Deploy new database
	echo "Deploying database";
	mysql -u$S_SETUP_DB_USERNAME -h$S_SETUP_DB_HOSTNAME --password="$S_SETUP_DB_PASSWORD" -e"CREATE DATABASE $S_PROMS_DB_NAME"
	mysql -u$S_SETUP_DB_USERNAME -h$S_SETUP_DB_HOSTNAME --password="$S_SETUP_DB_PASSWORD" -e"GRANT ALL PRIVILEGES ON $S_PROMS_DB_NAME.* TO $S_PROMS_DB_USERNAME IDENTIFIED BY '$S_PROMS_DB_PASSWORD' WITH GRANT OPTION"
	mysql -u$S_SETUP_DB_USERNAME -h$S_SETUP_DB_HOSTNAME --password="$S_SETUP_DB_PASSWORD" $S_PROMS_DB_NAME < sql/proms.sql
fi

#----------------------------------------------------------------------------
# Prepare codebase for new settings
#----------------------------------------------------------------------------
echo "Preparing settings.php";
REPLACE=""
REPLACE="${REPLACE}s|^define (\"REF001_.*|define (\"PROMS_VERSION\", \"$VERSION\");|;"
REPLACE="${REPLACE}s|^define (\"REF002_.*|define (\"PROMS_EMAIL\", \"$S_PROMS_LIST_EMAIL\");|;"
REPLACE="${REPLACE}s|^define (\"REF003_.*|define (\"DB_DATABASE\", \"$S_PROMS_DB_NAME\");|;"
REPLACE="${REPLACE}s|^define (\"REF004_.*|define (\"DB_USERNAME\", \"$S_PROMS_DB_USERNAME\");|;"
REPLACE="${REPLACE}s|^define (\"REF005_.*|define (\"DB_PASSWORD\", \"$S_PROMS_DB_PASSWORD\");|;"
REPLACE="${REPLACE}s|^define (\"REF006_.*|define (\"DB_HOSTNAME\", \"$S_PROMS_DB_HOSTNAME\");|;"
REPLACE="${REPLACE}s|^define (\"REF007_.*|define (\"SMTP_HOSTNAME\", \"$S_PROMS_SMTP_HOSTNAME\");|;"
REPLACE="${REPLACE}s|^define (\"REF008_.*|define (\"SMTP_PORT\", \"$S_PROMS_SMTP_PORT\");|;"
cat src/settings.php | sed "${REPLACE}" > temp.settings.php

#----------------------------------------------------------------------------
# Deploy codebase
#----------------------------------------------------------------------------
echo "Deploying codebase";
mkdir -p $S_SETUP_PATH
cp -R src/* $S_SETUP_PATH
mv temp.settings.php $S_SETUP_PATH/settings.php

if [ "$HAS_ROOT" == "1" ]; then
	echo "Setting rights";
	chown $S_SETUP_WEB_USERNAME:$S_SETUP_WEB_GROUPNAME $S_SETUP_PATH -R
	chmod 550 $S_SETUP_PATH 
	chmod 640 $S_SETUP_PATH/*.php
	chmod 550 $S_SETUP_PATH/images
	chmod 750 $S_SETUP_PATH/files
fi

#----------------------------------------------------------------------------
# Clean up
#----------------------------------------------------------------------------

output_text "Done" "The installation/upgrade of PROMS is done. If this is the first time you have installed PROMS, you must point your webbrowser to whatever directory it is that you installed PROMS in and use the 'Signup' feature to create a new account for yourself. The first account you create will be the superuser of PROMS. After that you can begin to add new projects, etc."
output_text "Notices" "Please make sure you have have Magic Quotes enabled in your installation of PHP. If you do not have this enabled, your installation of PROMS will be VULNERABLE to SQL INJECTIONS!\n\nIf you are running a different web server than Apache, please use your web server's configuration to prevent people from uploading scripts to the files/ directory! If you do not, people might be able to upload PHP or PERL files and execute from within PROMS, which opens up major security holes.\n\n"
